### PR TITLE
Fixes #6775 Error scalar function

### PR DESCRIPTION
**Pull request by ozdemircs**
This PR adds the Error debug function similar to the [BigQuery](https://cloud.google.com/bigquery/docs/reference/standard-sql/debugging_functions) as discussed in the #6775 and here.

### CODE REVIEW

File: src/function/scalar/generic/assert.cpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                                            vector<unique_ptr<Expression>> &arguments) {
+	bound_function.return_type = arguments[0]->return_type;
+	return nullptr;
+}
+
+void AssertFun::RegisterFunction(BuiltinFunctions &set) {
+	ScalarFunctionSet assert_function_set("assert");
+	auto assert_fun =
+	    ScalarFunction({LogicalType::ANY, LogicalType::BOOLEAN}, LogicalType::ANY, AssertFunction, AssertBindFunction);
+	assert_fun.null_handling = FunctionNullHandling::SPECIAL_HANDLING;
```
Comment (Mytherin) [inline src/function/scalar/generic/assert.cpp:44]: This function should be labeled as `FunctionSideEffects::HAS_SIDE_EFFECTS`, which the optimizer can then use to limit optimizations around the function.

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! After reading the discussion on #6775 I prefer the `ERROR` option for three reasons, (1) the ability for the user to specify an error message, (2) compatibility with BigQuery, (3) simplicity and clarity in what the function accomplishes. It is rather unusual for `assert` to return its input value, and it took me a while to parse what was actually happening here. I think the `ERROR` function is a cleaner solution that would be more understandable to users.

Apologies for not realizing the discussion happening there earlier.

**Comment by Tishj** [comment]
Ah my bad for suggesting it then.

Though I think the reasoning was that if the result is unused, it could get optimized out.

That is what inspired the `value.assert(x -> x != 0)` idea

**Comment by Mytherin** [comment]
In the [BigQuery example](https://cloud.google.com/bigquery/docs/reference/standard-sql/debugging_functions) they avoid the optimizer by either using the function in the `WHERE` clause with an `IF` condition, or by using it in a `CASE` statement. I think that is fine as a work-around - it just seems more readable to me.

> value.assert(x -> x != 0)

The issue with this syntax is that it is very geared towards this one case (checking for 0 on a single variable) - it would not work in general if you have more complex conditions you want to check. I also think the `ERROR` function has potential other purposes as a debug function because it can be used to e.g. test correct error handling in clients.

File: src/function/scalar/generic/assert.cpp (hunk 1 above)
Comment (ozdemircs) [reply src/function/scalar/generic/assert.cpp]: Ack. Updated the code. Please see the changes.

**Comment by ozdemircs** [comment]
Thanks for the review @Mytherin, I updated the description of the  current PR and code in the same branch, if you need me, to close this PR and send a new one please let me know.

I overall agree with the concerns. I think, we made the problem a bit complicated.

@Tishj thanks for the help and thinking together in the discussion, no worries, the purpose was to get familiar with the code, and the assert was a good start.

**Comment by Mytherin** [comment]
Thanks for the fixes! LGTM

**Pull request merged**